[envoy/admin] Scaffolding for structured admin and config_dump endpoint

jsedgwick opened the pull request:
See https://github.com/envoyproxy/envoy/pull/2771 for context

I will add documentation once the interfaces and placement of these things is settled. Current location (new admin package) was agreed upon as a good start with @htuch but i don't feel strongly about that or naming.


Signed-off-by: James Sedgwick <jsedgwick@lyft.com>

htuch commented on envoy/admin/admin.proto:
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
```
This should be `envoy/admin/v2/admin.proto` as per versioning elsewhere in data-plane-api.

jsedgwick replied:
Kk. So basically the whole repo is v2?

htuch replied:
At this point, yep. In the future we may have independent subtrees evolve at different rates, but this is probably medium-long term.

htuch commented on envoy/admin/admin.proto:
```diff
@@ -0,0 +1,17 @@
+import "gogoproto/gogo.proto";
+
+message ConfigDump {
+  map<string, google.protobuf.Any> configs = 1 [(gogoproto.nullable) = false];
```
Can you add comments? It's hard to know what this very general type is representing.

mattklein123 replied:
+1. In general, we should add full comments here for which admin endpoints these map to, how they are used, etc. It's not completely clear to me whether we want strongly typed output (maybe we do) vs. something general like this type.

jsedgwick commented:
Moved to v2/ and added some comments.

Btw, I tried to put in the protodoc `:ref:`s but I don't think I had the format right. Besides, not sure where we would want these hooked in. Is our doc generator setup documented anywhere besides what's in `CONTRIBUTING.md` now?

mattklein123 commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
```
Can we move this to `config_dump.proto`? I think we should have a discrete proto file for each endpoint in the admin package.

jsedgwick replied:
Yeah, agreed, was just using a catch-all to punt on naming. Will do.

mattklein123 commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+
+import "gogoproto/gogo.proto";
+
+// [#protodoc-title: ConfigDump]
```
I wouldn't worry about RST docs right now for this. There is a generally open item for how to document APIs in RST. Right now we don't really do it. We just document configuration. (In any case I don't think it's critical that these protos have HTML docs right now). We can figure it out in the context of https://github.com/envoyproxy/data-plane-api/issues/351.

jsedgwick replied:
Sounds good. I didn't realize we weren't generating docs for the APIs, good that's it's on the list.

mattklein123 commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+
+// The /config_dump admin endpoint uses this wrapper message to maintain and serve arbitrary
+// configuration information from any component in Envoy.
+message ConfigDump {
```
I don't think we need to do it right now, but in the future I think we want to make each endpoint a full RPC with a request and response message. This will allow us to serve both binary gRPC/REST and take parameters, etc. Can you potentially add a TODO here to that effect?

jsedgwick replied:
👍 Yup

htuch commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+  // routing subsystem might use "routes" as the key for its config, for which it uses the message
+  // RouteConfigDump as defined below. In the future, the key will also be used to filter the output
+  // of the /config_dump endpoint.
+  map<string, google.protobuf.Any> configs = 1 [(gogoproto.nullable) = false];
```
@louiscryan advocated that we maintain these in a format suitable for replay. For that to work, I think the string here should be the type URL. We need to think how this is going to work with https://github.com/envoyproxy/data-plane-api/pull/527. We should have a doc for this by later this week. I think we might want this to be something like `type URL -> resource name -> (resource proto x version)` or something like this. It depends what we settle on for versioning in incremental xDS.

mattklein123 replied:
Yeah +1 it would be nice to just have version be part of this, then we can also solve https://github.com/envoyproxy/envoy/issues/2172.

jsedgwick replied:
I originally had that but saw that Any has the `type_url`, as in https://gist.github.com/jsedgwick/3a46408a9728ccef9a63d32b4c463c2b
Now, it's missing the version - is that a deficiency of Any?
So currently it's more like `resource name -> type URL -> resource proto'. Does it make more sense to group `type_url` and the actual resource, just like with Any? The keep the resource name top level and bundle in the version at the end.

Anyways, bikeshedding, we can just hold on that until you guys settle on versioning.

mattklein123 replied:
You don't need to wait on the versioning. Just update the map target to be a struct which contains a string version and an Any object.

jsedgwick replied:
Hold on, what exactly is the version we're encoding? Or is that what's pending resolution? Shall I just fill the field with empty strings?

mattklein123 replied:
No, you are filling it with https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/discovery.proto#L60

mattklein123 replied:
Which will resolve envoyproxy/envoy#2172

jsedgwick replied:
I see. But in this case, those belong to `RouteConfiguration`, not `RouteConfigDump`, no? Which version would be pair with the `RouteConfigDump` in the `Any`?

mattklein123 replied:
RDS is a bit of a tricky one, since technically without ADS there can be multiple RDS streams, 1 per HTTP connection manager. In that case, yes, the version likely needs to go with the route configuration. For the other cases it could go at the top level (CDS, LDS, etc.).

Another way of doing it, probably better, would be to track the discovery response version that carried a discrete cluster, route, listener, etc. and print that along with each object. This would work for incremental updates better. So maybe that's that we should do. In light of that I would take a look at the message structure and figure out a nesting that works. I don't think you need to wait until the incremental stuff is decided.

jsedgwick replied:
🆗

htuch replied:
@mattklein123 One of the possible incremental xDS option is to have a version at resource level. Let's discuss there and circle back here.

mattklein123 replied:
@htuch OK SGTM.

htuch commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+message ConfigDump {
+  // This map is serialized and dumped in its entirety at the /config_dump endpoint.
+  //
+  // Keys should be a short descriptor of the config object they map to. For example, envoy's HTTP
```
Nit: s/envoy/Envoy/

jsedgwick replied:
:+1:

htuch commented on envoy/admin/v2/admin.proto:
```diff
@@ -0,0 +1,31 @@
+// their RouteConnfiguration objects. Static routes configured in the bootstrap configuration are
+// separated from those configured dynamically via RDS. This message is available at the
+// /config_dump admin endpoint.
+message RouteConfigDump {
```
This is a good point; we want to be able to dump static and dynamic information. What if we have an endpoint to dump the entire bootstrap? Would that be sufficient for covering the static requirements?

mattklein123 replied:
Yes we should definitely have a bootstrap dump option. I could go either way on whether that covers though, because for example a route table can come as part of a static LDS route. I think it's probably OK for "routes" to dump both static and dynamic.

jsedgwick replied:
Yes, the Bootstrap should definitely have a dump. Is that the sort of thing I should be opening issues for? I'm keeping notes locally so I don't forget.

More to the point, it seems helpful to have this distinction available without going for the bootstrap dump. IMO RouteConfiguration could use a flag indicating its static/dynamic origin but maybe that ship has sailed.

mattklein123 commented:
@jsedgwick my advice is to mark this proto as "draft" type (see other examples) and go ahead and finish up the envoy side PR. We can add in version info next as we are saying the proto is not finalized.

htuch commented on envoy/admin/v2/config_dump.proto:
```diff
@@ -0,0 +1,35 @@
+  // This map is serialized and dumped in its entirety at the /config_dump endpoint.
+  //
+  // Keys should be a short descriptor of the config object they map to. For example, envoy's HTTP
+  // routing subsystem might use "routes" as the key for its config, for which it uses the message
```
I think we should consider making this a type URL in the future. I'm still after a config that could be effectively replayed via an xDS management server.
